### ISSUE TITLE
Rack 2.0 Protocol Proposal (discussion)

### ISSUE THREAD

**Issue by boazsegev** [rack/rack#1707]
This is my latest vision for a Rack 2.0 protocol specification and I think it would actually make things a lot easier.

It offers server agnostic extendability, making it easy to test for server features and add whatever might be missing using external gems.

It removes many of the biases that the CGI model has against evented servers while also being easily implementable by blocking servers (process/thread/fiber per connection models). 

Backwards compatibility can (should) be easily implemented as an extension, so it allows for a gradual transition.

It makes polling / streaming / slow responses into first class citizens, moving some of the weight off `hijack` into a simpler API.

I called it NeoRack because, well, why not.

I would love to read feedback and learn if the Rack team would consider this approach...?

Thank you for your time.

P.S., sorry about the length of it, but it covers more ground and includes code examples.

---

# NeoRack Specification

The following are the NeoRack Application, Server and Extension specifications that NeoRack Applications, Servers and Extensions **MUST** follow in order to be considered conforming.

Note that NeoRack Applications, Servers and Extensions are **NOT** required to include any specific NeoRack related gem in order to conform to these specifications.

## NeoRack Applications

A NeoRack application is any Ruby object that responds to the method `call` where the `call` method accepts two arguments: `request` and `response`. i.e.,

```ruby
APP = Proc.new {|request, response| response << "Hello World" }
```

If an Application doesn't call neither `response.stream` nor `response.finish` before it returns, the Server should call `response.finish`.

i.e., in pseudo code:

```ruby
APP.call(request, response)
response.finish unless response.streaming? || response.finished?
```

For backwards compatibility reasons, NeoRack Applications **MUST NOT** return an Array as their final value, as that may cause the Server to assume that these are old-Rack Applications and process them in a backwards compatible manner.

## The NeoRack Request Object

The NeoRack `request` Object **SHOULD** be an instance of a class that inherits from Hash (but is **NOT** an instance of Hash) and **MUST** implement the following "Hash-like" methods: `[]`, `[]=`, `each`, `size`, `has_key?` and `merge!` in the same way they are implemented by Hash.

**Note**: during Server initialization, an application might `extend` or `include` its own modules into the Server's Request class. For this reason it is better if that class is not Hash class itself.

### Request methods

The `request` object **MUST** respond to the following methods:

* `server` - returns the Server class / object that called the Application object.

### Required Request key-value pairs

The following key value pairs **MUST** be set by the Server before forwarding the `request` object to the Application:

* `:VERSION_SPEC` (Symbol) - the version for this specification as a three member Array of Numbers, currently `[0,0,1]`.

* `:VERSION`      (Symbol) - the HTTP version String as reported by the client. For HTTP/2 and HTTP/3 (QUIC) use `'HTTP/2'` and `'HTTP/3'` respectively.

* `:secure`       (Symbol) - **MUST** be set to false unless the server itself handles TLS/SSL encryption for this request / connection.

* `:scheme`       (Symbol) - depending on the request URL and environment, usually `http`, `ws`, `https` or `wss`.

* `:method`       (Symbol) - the HTTP method used for the request, set as a String instance (`'HEAD'`, `'GET'`, `'POST'`, `'UPDATE'`, etc').

* `:query`        (Symbol) - The portion of the request URL that follows the `'?'`, if any. **MUST** be a String (even if empty).

* `:path`         (Symbol) - The portion of the request URL before the query and after the host name and optional authentication details.

    **MUST** be a String. **MUST** start with the `'/'` character (even if the original path was an empty String).

    Servers that support multiple Application objects **MAY** change this value in a documented manner.

* `:path_root`        (Symbol) - The portion of the request URL path that was extracted from the beginning of the `:path` String or an empty String.

    **MUST** be a String (even if empty).

    **MUST NOT** end with `'/'`.

    Unless empty, **MUST** start with a `'/'`.

    Servers that support multiple Application objects **SHOULD** document how `path_root` is used.

**Note**: it **MUST** be possible to construct a valid URL that routes to the same Application object using:

```ruby
url = "#{request[:scheme]}://#{request['host']}#{request[:path_root]}#{request[:path]}?#{request[:query]}"
```

### Optional Request key-value pairs

* `:body` (Symbol) - an HTTP payload (body) object (see details further on). If a payload exists, this value **MUST** be set. Otherwise this **MAY** be set to `nil`, left unset or point to an empty body object.

### HTTP Request Headers

All HTTP headers, except `'content-length'` and `'content-type'`, **MUST** be set as key-value pairs in the `request` where the key is a **String** and the value is either a String or an Array of Strings.

The HTTP headers `'content-length'` and `'content-type'` **MAY** be set as key-value pair in the `request` object, but doing so isn't required. If set, the value of the key `'content-length'` **MUST** be a String, as all header values **MUST** be either Strings or Arrays.

Header names **MUST** be converted to their lower case equivalent before the key-value pair is set (i.e., `"Content-Length"` MUST be converted to `"content-length"`).

All non-header key-value pairs **SHOULD** use Symbol or Number keys.

Unless the Server implements the `:backwards_compatible` extension, HTTP headers **MUST** be the **ONLY** String keys in the `request` set by the Server or by any Extension. Otherwise, it might make it impossible for a `:backwards_compatible` extension to be implemented externally.

If a header arrives only once (the common case), it's value **MUST** be set as a String instance object.

If a header arrives more than once, (i.e., `Cache-Control`) it **SHOULD** be set as an Array instance object with each value (by order of arrival) inserted as a String instance member into the Array. Such an Array **MUST ONLY** contain String instance objects (i.e., `headers['cache-control'] = ['no-cache', 'no-store']`).

**ONLY IF** [the HTTP protocols allows for this variation](https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2), Servers **MAY** also merge the header values into a single comma separated value list String instance.

Servers **MAY** also split comma separated lists in header values into Arrays of Strings.

Applications **MUST** be prepared to handle both variations (Array / String).

#### HTTP Request Body

The NeoRack Request Body (`request[:body]`) is an IO-like object which contains the raw HTTP POST data (payload / body). It is backwards compatible with the original Rack specification with the addition of two methods: `length` and `type`.

When no HTTP payload / body was received, `request[:body]` **SHOULD** be `nil`, but **MAY** be a Request Body Object that maps to an empty String.

The `request[:body]` object (if set) **MUST** respond to IO style methods `gets`, `each`, `read`, `rewind` and `length`. These methods **MUST** behave the same way as described by [the Ruby IO documentation](https://ruby-doc.org/core-2.7.1/IO.html). The following clarifications were pretty much copied from [the original Rack specification](https://github.com/rack/rack/blob/master/SPEC.rdoc):

* `gets` **MUST** be called without arguments and Reads the next “line'' from the I/O stream. It returns either a String (on success), or `nil` (on `EOF`).

* `read` behaves like `IO#read`. Its signature is `read([length, [buffer]])`.

    If given, length **MUST** be a non-negative Integer (>= 0) or `nil`,

    If length is given and not `nil`, then this method reads at most length bytes from the input stream.

    If length is not given or `nil`, then this method reads all data until `EOF`.

    When `EOF` is reached, this method returns `nil` if length is given and not `nil`, or “” if length is not given or is `nil`.

    If given, buffer **MUST** be a String (never `nil`).

    If buffer is given, then the read data will be placed into buffer instead of a newly created String object.

* `each` **MUST** be called without arguments and only yield one String instance object per iteration.

* `rewind` **MUST** be called without arguments. It rewinds the input stream back to the beginning. It **MUST NOT** raise `Errno::ESPIPE`: that is, it may not be a pipe or a socket. Therefore, developers **MUST** buffer the input data into some rewindable object if the underlying input stream is not rewindable.

* `close` **MUST NOT** be called on the body object except by the Server.

The `request[:body]` object (if not `nil`) **MUST** also implement the following (non IO related) methods:

* `type` returns the String value of the HTTP `content-type` header (if received).

**Note**: Applications **SHOULD NOT** access the `'content-type'` or `'content-length'` headers using the request object and **SHOULD** prefer to access this data using the `request[:body]` object (`request[:body].type` and `request[:body].length`).

## The NeoRack Response Object

The NeoRack Response Object (`response`) manages the response data, including the response status, header data and body (payload).

### Timing Considerations

The response status and the header header data **MUST** remain mutable until either `response.stream` is called, `response.finish` is called, or the Application returns (`app.call(request, response)` returns).

At the earliest of these events (`response.stream` is called, `response.finish` is called, or the Application returns):

* The Server **MUST** prepare to send the status and header data to the client.

* The Server **MUST** also prepare to send the body of the response (if required, i.e., request wasn't `HEAD`) and either:

    * calculate the `'content-length'` header (non streaming-response); or

    * use the proper encoding and update the proper `'transfer-encoding'` header (for streaming responses);

    These actions are subject to the HTTP protocol requirements. There may be different protocol requirements for different HTTP versions.

Once `response.stream` was called, any attempt to write to the response (see `write` or `<<`) **SHOULD** schedule the data to be sent as required by the protocol and as implemented by the Server.

Note that Servers **MAY** block to until the data was actually sent through the socket.

Once `response.finish` was called, any attempt to write to the response **MUST** be either ignored or result in an exception, as specified below.

### NeoRack Response Methods

The NeoRack Response Object **MUST** respond to the following methods: 

#### `status`

Returns the response status.

#### `status=`

Sets the response status. **MAY** be called multiple times before the Application returns.

**SHOULD** raise an exception if either `streaming?` or `finished?` would have returned `true`.

#### `add_header(name, value)`

Adds a header to the response. If the header exists, the header will appear multiple times.

If either `name` or `value` are `nil`, does nothing (returns, no exception is raised).

**MUST** raise an exception if either `streaming?` or `finished?` would have returned `true`.

`name` and `value` **MUST** be either `nil`, a String object or a Number. Otherwise an `ArgumentError` exception **MUST** be raised.

The header `name` **SHOULD** be lowercase, but Servers **SHOULD** expect application to be inconsistent in this regard.

The Server **MAY** perform any additional action, such as sending an Early Hints responses or invoking an HTTP push promise request for `'link'` headers, etc'. Such behavior **SHOULD** be documented.

#### `reset_header(name, value = nil)`

If `name` is `nil`, does nothing (returns, no exception is raised).

If `value` is `nil`, deletes the header from, the response if the header exists.

If both `name` and `value` are set, adds a header to the response. If the header exists, it is overwritten.

**MUST** raise an exception if either `streaming?` or `finished?` would have returned `true`.

`name` and `value` **MUST** be either `nil` or a String object. Otherwise an `ArgumentError` exception **MUST** be raised.

The header `name` **SHOULD** be lowercase, but Servers **SHOULD** expect application to be inconsistent in this regard.

The Server **MAY** perform any additional action, such as sending an Early Hints responses or invoking an HTTP push promise request for `'link'` headers, etc'. Such behavior **SHOULD** be documented.

#### `set_cookie(name, value, options={})`

If `name` is `nil`, does nothing (returns, no exception is raised).

If `value` is `nil`, sets the cookie to be deleted.

If both `name` and `value` are set, adds a cookie to the response. If the cookie exists, it is overwritten.

**MUST** raise an exception if either `streaming?` or `finished?` would have returned `true`.

Servers **SHOULD NOT** encode cookie data (names / value). Encoding the data is the Application's responsibility and choice.

`name` and `value` **MUST** be either `nil` or a String object. Otherwise an `ArgumentError` exception **MUST** be raised. **Note**: empty String objects are valid, but **SHOULD** be avoided by the Applications.

If `name` is an invalid cookie name or `value` is an invalid cookie value, behavior is undefined (see [here](https://tools.ietf.org/html/rfc6265) and [here](https://stackoverflow.com/questions/1969232/what-are-allowed-characters-in-cookies)). The Server **MAY** attempt to auto-correct the issue, raise and exception, quietly fail, etc'.

The `option` hash **MUST** recognize the following Symbols for setting cookies (see also the [`Set-Cookie` header details](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie)):

* `:path`       - A path that must exist in the requested URL, or the browser won't send the Cookie header.

* `:domain`     - The domain for which this cookie applies. Valid a specific domain String object (i.e. `'example.com'`). Note that subdomains are always allowed.

* `:same_site`  - If set, **MUST** be either `:strict`, `:lax` or `:none`. [Read details here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie).

* `:secure`     - A secure cookie is only sent to the server when a request is made with the `https` scheme.

* `:http_only`  - if `true`, forbids JavaScript from accessing the cookie through a script (i.e, using `document.cookie`).

   **Note**: The cookie is still sent with JavaScript originated requests. 

* `:max_age`    - The number of seconds until the cookie expires. A zero or negative number will expire the cookie immediately.

* `:expires`    - **SHOULD** be considered deprecated by Application developers. It is only provided for backwards compatibility.

    `:expires` is the maximum "due date" lifetime for the cookie as either a number (Unix style time stamp) or a Ruby Time instance object. If provided, servers **MUST** translate the time stamp into an HTTP Date format.

    If both `:max_age` and `:expires` are set, `:max_age` **MUST** be preferred and `:expires` **SHOULD** be ignored. Servers **MAY** send both values (`Max-Age` always has precedence).

    If only `:expires` is set, the Server **MAY** convert `:expires` to `:max_age` values.

#### `write(data, offset = 0, length = nil)`

Adds an object `data` to the response payload (body).

Returns `self` (the `response` object).

**MUST** raise an exception if `finished?` would have returned `true`.

`data` **MUST** be either:

* Be a String instance object.

* Respond to both `fileno` and `close`, allowing the Server to take ownership of the IO device and stream it.

If `offset` is set, the first `offset` bytes in the `data` object are ignored (not sent). `offset` **MUST** be a Number.

If `length` is set, up to `length` bytes from the `data` object will be sent. `length` **MUST** be either `nil` or a Number.

Servers **MUST** test the `length` value for overflow whenever possible. If `length` overflows it is **NOT** an error, but it could cause network errors that the Server **MUST** properly handle.

If `streaming?` would return `true`, the Server **SHOULD NOT** wait for `response.finish` or for the Application to return before sending the data.

If `streaming?` would return `false`, the Server **SHOULD** wait for `response.finish` to be called, for `response.stream` to be called, or for the Application to return, before sending the data.

#### `<<`

The `<<` method is an alias to `write`.

#### `stream`

Starts streaming the response. If `request[:method] == 'HEAD'`, the method call **MUST** call `finish` and return instead.

Returns `self` (the `response` object).

Once an Application calls `stream` and until the Application calls `finish`, the method `streaming?` **MUST** return `true`.

Servers **MUST** prepare to send the status, headers and any previously scheduled data at this point.

**MUST** raise an exception if `finished?` would have returned `true`.

#### `streaming?`

**MUST** return `false` unless `stream` was called and `finish` wasn't called.

Otherwise, **MUST** return `true`.

#### `finish`

Finalizes the response.

**SHOULD** raise an exception if `finished?` would have returned `true`, but **MAY** quietly fail without doing anything.

The Server **MUST** prepare to send all headers and data at this point.

The Server **MUST** call (or schedule) all the `run_after` callbacks at this point.

#### `finished?`

**MUST** return `false` **unless** either `finish` or `cancel` were called.

Otherwise, **MUST** return `true`.

#### `cancel(error_code)`

Cancels the response object, returning the response management to the Server.

If possible, the Server **MUST** send an appropriate error response.

**SHOULD** raise an exception if `error_code` is less than 400 or grater than 599 (`< 400 || >= 600`).

**MUST** raise an exception if either `streaming?` or `finished?` would have returned `true`.

## NeoRack Servers

A NeoRack Server **MUST**, at the very least, handle network and HTTP protocol details to the minimal level required to implement the NeoRack `request` and  NeoRack `response` objects and call a NeoRack Application object.

A NeoRack Server **MUST** accept one or more NeoRack Applications and forward a `request` and `response` object to the Application.

If a NeoRack Server accepts more than a single NeoRack Application, the Server **SHOULD** document how it selects the Application object.

### Server methods

A NeoRack Server **MUST** implement the following methods in the Server object returned by a call to `request.server`:

* `forked?`  - **MUST** be set to `true` **ONLY IF** the Server's concurrency model expects to use `fork` (i.e., workers processes or a process-per-connection). If this is not the case, returns `false`.

* `blocking?` - **MUST** be set to `true` **ONLY IF** the Server's concurrency model allows execution to block without negavive side-effects, i.e., process/thread/fiber-per-connection design patterns.

* `extensions`  - **MUST** be a Hash object that maps extension name Symbols to version number arrays. **MAY** be an empty Hash.

    Version number arrays **MUST** be frozen Array who's first three members are Numbers that follow [semantic versioning](https://semver.org).

    i.e.: `request[:SERVER].extensions[:rack_compatible] # => [1,3,0]`

* `classes`  - **MUST** be a Hash object where Symbol keys are mapped to Class objects (not Class instances). **MAY** be an empty Hash.

    The key `:request` **MUST** map to the Class used for the Server's `request` objects.

    The key `:response` **MUST** map to the Class used for the Server's `response` objects.

    The key `:concurrency` **MUST** map to either `Thread` or `Fiber`, allowing the Server to hint at the preferred way for the Application to handle concurrent processing. Servers **SHOULD** set this value to `Thread` unless they are `Fiber` scheduling aware.

    Extensions to this specification **MAY** require additional names to be added as long as these names are unique.

    This **MAY** be used to test for and implement extensions. i.e.:

    ```ruby
    # an Application could call this or a similar line using the DSL (see later) to add features to the Response object
    server.classes(:response).include MyNeoRackHelpers unless server.extensions[:common_helpers] && server.extensions[:common_helpers][0] == 1
    ```

### Server CLI (Command Line Interface)

A NeoRack Server **MUST** document its startup and shutdown procedures.

A NeoRack Server **MAY** offer a Ruby API for setting up and/or running and/or managing the Server from within Ruby.

A NeoRack Server **SHOULD** implement a CLI (Command Line Interface) and **SHOULD** expose as many options through its CLI, minimizing the need for setup files / code.

A NeoRack Server **SHOULD** accept at least one file name as an unnamed CLI argument and execute that file name as a Ruby Script within a context that allows the Script to access the Server's NeoRack DLS. The default script name is `config.ru`.

The following common argument names are **RECOMMENDED** (but to each their own):

* `-b` - the address to listen to.

* `-p` - the port number to listen to.

* `-D` - log Debug messages, if any.

* `-url` - possible alternative to the `-b` and `-p`, allowing a URL type address for the address and port binding.

* `-l` - log HTTP requests once the response was sent (finished), if supported.

* `-k` - sets HTTP keep-alive timeout in seconds.

* `-maxbd` - sets the approximate HTTP upload limit in Mb.

* `-maxhd` - sets the approximate total header length limit per HTTP request in Kb.

* `-t` - if multi-threaded mode is supported, sets the number of threads to be used.

* `-w` - if a process worker pool is supported, sets the number of worker processes to be used.

It is **RECOMMENDED** that Servers automatically test for the `ADDRESS` and `PORT` environment variables if those aren't set by the command line.

### NeoRack Server DSL

The following method must be made available to the script as if they were global methods (DSL):

* `server` - returns the Server object, allowing access to the Server methods.

* `run(application)`  - set the NeoRack Application object for the current Script.

* `run_before(proc_obj = nil, &block)` - Either the `proc_obj` or `block` **MUST** respond to `.call(request, response)` and will be called **before** the Application starts processing the request. This may be used for pre-request logic, such as authentication, database connection checkout, etc'.

* `run_after(proc_obj = nil, &block)` - This object **MUST** respond to `.call(request)` and will be called **after** each response ended (handled along with `response.finish`). This may be used for cleanup logic, such as removing database connections from the `request` Hash, logging, etc'.

* `use(middleware, *args, &block)` - provided for for backwards compatibility with nested middleware design pattern.

* `warmup(proc_obj = nil, &block)` - provided for for backwards compatibility, takes a block and/or Proc object that will be called with the final Application object.

i.e.:

```ruby
module NeoRack
  # A single application NeoRack application DSL script handler 
  #
  # Internal names end with `___` to minimize possible namespace concerns.
  class Builder
    # initialize a new builder object and run the script in filename
    def initialize(server, filename = 'config.ru')
      @server___, @app___, @warmup___, @app___= server, , nil, nil, nil
      @stack_pre___, @stack___, @stack_post___  = [].dup, [].dup, [].dup
      script = ::File.read(filename)
      # remove UTF-8 BOM, see: https://stackoverflow.com/questions/2223882/whats-the-difference-between-utf-8-and-utf-8-without-bom
      script.slice!(0..2) if script.encoding == Encoding::UTF_8 && script.start_with?('\xef\xbb\xbf')
      # run script in context of the object, enabling the DLS
      instance_eval(script)
    end

    # DSL method - access the Server object and its methods
    def server
      @server___
    end

    # DSL method - set the application to be used by the Script
    def run(application)
      # add middleware to a middleware stack
      @app___ = application
    end

    # DSL method - runs `.call(request, response)` before the application handles the response.
    #
    # Used pre-request logic, such as authentication, database connection checkout, etc'.
    def run_before(prc = nil, &block)
      prc ||= block
      raise(ArgumentError, "this method requires an object that responds to `call(request, response)`") unless(prc.respond_to?(:call))
      @stack_pre___ << prc
    end

    # DSL method - runs `.call(request)` after the response ended (when steaming, this is delayed until streaming ends).
    #
    # Used for cleanup logic, such as removing database connections from the `request` Hash, logging, etc'.
    def run_after(prc = nil, &block)
      prc ||= block
      raise(ArgumentError, "this method requires an object that responds to `call(request, response)`") unless(prc.respond_to?(:call))
      @stack_pre___ << prc
    end

    # DSL method for backwards compatibility 
    def use(middleware, *args, &block)
      # add middleware to a middleware stack
      @stack___ << [middleware, args, block]
    end

    # DSL method for backwards compatibility 
    def warmup(prc = nil, &block)
      @warmup___ ||= prc || block
    end

    # Internal use: returns the setup callback stack, the application object and the cleanup callback stack.
    def build_stack___
      raise "Application object missing!" unless @app___
      @stack___ << @app___
      app = @stack___.pop
      while(tmp = @stack___.pop)
        if tmp[3]
          app = tmp[0].new(app, *tmp[1], &tmp[2])
        else
          app = tmp[0].new(app, *tmp[1])
        end
      end
      @app___ = app

      @warmup___.call(@app___) if @warmup___
      @stack_pre___, @app___, @stack_post___
    end

    # returns the setup callback stack, the application object and the cleanup callback stack.
    def self.load(server_klass, filename)
      instance = NeoRack::Builder.new(server_klass, filename)
      instance.build_stack___
    end
  end
end
```

## NeoRack Extensions

NeoRack Servers cab be extended either internally, through the Server supporting an extension, or externally, through setting up middleware and/or adding mixins to the published classes and updating the Aerver's `extensions` and `classes`.

Extensions registered in the NeoRack repository will have their specifications published in the `extensions` folder and **MUST** specify a unique name for the Servers `extensions` Hash, specify a [semantic versioning](https://semver.org) compliant 3 numbered version array and be mature.

Developers may also register extension drafts to be placed in the `extensions/drafts` folder and promote community feedback.

### External Extensions

External extensions **SHOULD** implement a `register` method that accepts at least one argument - the `self` object of the DSL execution environment.

The following is a made-up logging extension implementation example that should be considered as pseudo code:

```ruby
# Place module somewhere
module MyLoggingExtension
  # registers the extension from the DLS
  def self.register(dsl)
    return if server.extensions[:logging_example]
    dsl.server.extensions[:logging_example] = [0,0,1]
    dsl.run_before(self.method :on_start)
    dsl.run_after(self.method :on_finish)
  end
  # Marks time
  def self.on_start(request, response)
    request[:STARTED_AT] = Time.now
  end
  # Prints log
  def self.on_finish(request)
    delta = Time.now - request[:STARTED_AT]
    puts "HTTP request %s took %.4f"%(#{request[:path]}, delta)
  end
end

# in the DSL, require the module and perform:
MyLoggingExtension.register(self)
```

## NeoRack Compatibility with Rack

NeoRack backwards compatibility with [the CGI style Rack specifications](https://github.com/rack/rack/blob/master/SPEC.rdoc) is considered an Extension and uses the reserved extension name `:backwards_compatible`.

Backwards compatibility extensions **MUST**:

* Set all `env` values as set in the [CGI style Rack applications](https://github.com/rack/rack/blob/master/SPEC.rdoc).

* Set the Server's `extensions[:backwards_compatible]` values to the Rack protocol version they support (i.e., `[1,3,0].freeze`).

* Set `env['neorack.request']` to the `request` object.

* Set `env['neorack.response']` to the `response` object.

Backwards compatibility extensions **SHOULD** use the `request` object for implementing the Rack style `env`.

The following is an example for an external backwards compatibility extension, it is untested, might not work and should be considered pseudo code:

```ruby
# place this extension code somewhere
module NeoRack
  class BackwardsCompatibility
    def self.register(dsl)
      unless server.extensions[:backwards_compatible] && server.extensions[:backwards_compatible][0] == 1
        dsl.server.extensions[:backwards_compatible] = [1,3,0].freeze
        dsl.use(self)
      end
    end
    def initialize(app)
      @app = app
    end

    def call(request, response)
      add_old_env_values_to_request(request, response)
      old = @app.call(request)
      process_old_return_value(old, response)
    end

    # example implementation
    def self.add_old_env_values_to_request(request, response)
      new_headers = {}
      request.each do |k,v|
        if (k.is_a?(String) && k[0].ord >= 'a'.ord && k[0].ord <= 'a'.ord)
          if k == 'content-length' || k == 'content-type'
            new_headers[k.swapcase.gsub!('-', '_')] ||= v
          else
            new_headers["#{HTTP_}#{k.swapcase.gsub('-', '_')}"] ||= v
          end
        end
      end
      request.merge! new_headers
      # set whatever Rack requires... i.e.:
      request['REQUEST_METHOD']   = request[:method]
      request['rack.url_scheme']  = request[:scheme]
      request['SERVER_NAME']      = request['host']
      request['SCRIPT_NAME']      = request[:path_root]
      request['PATH_INFO']        = request[:path]
      request['QUERY_STRING']     = request[:query]
      request['rack.version']     = [1,3,0]
      request['rack.errors']      = STDERR
      if request[:body]
        request['rack.input']       = request[:body]
        request['CONTENT_LENGTH']   = request[:body].length.to_s
        request['CONTENT_TYPE']   = request[:body].type if request[:body].type
      else
        request['rack.input']     = StringIO.new
      end
      # allow NeoRack aware apps access to these objects
      request['neorack.request']  = request
      request['neorack.response'] = response
    end

    # example implementation 
    def self.process_old_return_value(old, response)
      # do nothing if it was already done.
      return nil if(response.finished? || response.streaming?)
      raise "unexpected return value from application" unless (old.is_a?(Array) && old.length == 3)
      # set status
      response.status = old[0].to_i
      # copy headers to new response object
      old[1].each {|name, val| val = val.split("\n"); val.each {|v| response.add_header(name, v)} }
      # send body
      case old[2].class
      when String
        response << str
      when Array
        old[2].each {|str| response << str }
      else
        # start streaming the response
        response.stream
        # perform `each` in a new thread / fiber, as it may block the server
        if(old['neorack.request'][:SERVER].blocking?)
            old[2].each {|str| response << str }
            response.finish
        else
          old['neorack.request'][:SERVER].classes[:concurrency].new do
            old[2].each {|str| response << str }
            response.finish
          end
        end
      end
    end
  end
end

# run this code in the DSL
NeoRack::BackwardsCompatibility.register(self)
APP = Proc.new {|env| [200, {}, ["Hello World"]] }
run APP
```

**Comment by jeremyevans** [comment]
I think you should release neorack as a separate gem/SPEC.  That way web servers and web frameworks that want the request/response API can switch to using it.  Rack's biggest strength is the simplicity and flexibility of the `env => [status, headers, body]` API, and I don't think we should change that.

**Comment by boazsegev** [comment]
Hi @jeremyevans ,

Sure, as you wish.

I thought that since this will be compatible with the 1.3.0 protocol version, it could work.

I put a lot of thought on this one (though not enough editorial) and I plan to implement it with backwards compatibility for iodine 0.8.0 (so applications that return an Array get treated as Rack applications).

Cheers.

**Comment by boazsegev** [comment]
Done.

It might take time, but I hope [NeoRack]((https://github.com/boazsegev/neorack)) and Rack will inspire each other.

I don't think NeoRack can or should replace Rack. NeoRack isn't even meant to have a code base... but maybe NeoRack can augment Rack and make some tasks easier to code by minimizing the use-cases that require `hijack`.

**Comment by jeremyevans** [comment]
@boazsegev I apologize if my previous response seemed dismissive, as that was not my intention. Your spec seems well designed.  I just don't think it is good as a replacement for rack's spec.  I think it is better as an alternative for rack's spec.  However, other committers may feel differently.

**Comment by boazsegev** [comment]
@jeremyevans ,

No need to apologize. In your shoes I would probably have had the same reaction.

Rack is the bread and butter of so much of the Ruby community. I totally understand that things need to be as stable as possible.

**Comment by nateberkopec** [comment]
@boazsegev I'm pretty conservative on the Rack front as well, but I appreciate that this is a very thoughtful proposal. I'll read it in full and give my thoughts over at NeoRack.

**Comment by boazsegev** [comment]
@nateberkopec , thank you. Any input is appreciated.
